pkg/scheduler/flavorassigner/flavorassigner.go

```diff
@@ -356,38 +355,48 @@ func (a *Assignment) append(requests workload.Requests, psAssignment *PodSetAssi
 	a.LastState.LastTriedFlavorIdx = append(a.LastState.LastTriedFlavorIdx, flavorIdx)
 }
 
-// findFlavorForResourceGroup finds the flavor which can satisfy the resource
-// request, along with the information about resources that need to be borrowed.
+// findFlavorForPodSetResource finds the flavor which can satisfy the podSet request
+// for all resources in the same group as resName.
+// Returns the chosen flavor, along with the information about resources that need to be borrowed.
 // If the flavor cannot be immediately assigned, it returns a status with
 // reasons or failure.
-func (a *Assignment) findFlavorForResourceGroup(
+func (a *FlavorAssigner) findFlavorForPodSetResource(
 	log logr.Logger,
-	rg *cache.ResourceGroup,
+	psId int,
 	requests workload.Requests,
-	resourceFlavors map[kueue.ResourceFlavorReference]*kueue.ResourceFlavor,
-	cq *cache.ClusterQueue,
-	spec *corev1.PodSpec,
-	lastAssignment int) (ResourceAssignment, *Status) {
+	resName corev1.ResourceName,
+	assignmentUsage cache.FlavorResourceQuantities,
+) (ResourceAssignment, *Status) {
+	resourceGroup, found := a.cq.RGByResource[resName]
+	if !found {
+		return nil, &Status{
+			reasons: []string{fmt.Sprintf("resource %s unavailable in ClusterQueue", resName)},
+		}
+	}
+
 	status := &Status{}
-	requests = filterRequestedResources(requests, rg.CoveredResources)
+	requests = filterRequestedResources(requests, resourceGroup.CoveredResources)
+	podSpec := &a.wl.Obj.Spec.PodSets[psId].Template.Spec
 
 	var bestAssignment ResourceAssignment
 	bestAssignmentMode := NoFit
 
 	// We will only check against the flavors' labels for the resource.
-	selector := flavorSelector(spec, rg.LabelKeys)
-	flavorIdx := -1
-	for idx, flvQuotas := range rg.Flavors {
-		if features.Enabled(features.FlavorFungibility) && idx <= lastAssignment {
-			continue
-		}
-		flavor, exist := resourceFlavors[flvQuotas.Name]
+	selector := flavorSelector(podSpec, resourceGroup.LabelKeys)
+	assignedFlavorIdx := -1
+	idx := 0
+	if features.Enabled(features.FlavorFungibility) {
+		idx = a.wl.LastAssignment.LastTriedFlavorForPodSetResource(psId, resName) + 1
```

Comment:
nit: the `+ 1` here needs to correspond with the  `-1` returned in `workload.go`. This may require jumping to understand. 

I'm wondering if we could have a function `NextFlavorIdxToTry` that would encapsulate the logic (including the check for `features.Enabled(features.FlavorFungibility)`) and always return >= 0.

Reply:
Good idea

Reply:
thx